sdk/src/main/java/io/opentelemetry/sdk/tags/TagMapSdk.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.tags;
+
+import static io.opentelemetry.internal.Utils.checkNotNull;
+
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.tags.Tag;
+import io.opentelemetry.tags.TagKey;
+import io.opentelemetry.tags.TagMap;
+import io.opentelemetry.tags.TagMetadata;
+import io.opentelemetry.tags.TagValue;
+import io.opentelemetry.tags.unsafe.ContextUtils;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+
+@Immutable
+public class TagMapSdk implements TagMap {
```

Comment:
Suggestion: consider using `@AutoValue`, then you don't need to worry about `hashCode` or `equals`.

Reply:
I was trying to figure that out... I've never used `@AutoValue`.  It makes sense for classes that are Immutable POJO's, but I wasn't sure how to get that to work with the semantics we wanted here.  Any examples I should follow?

Reply:
As an example, feel free to take a look at https://github.com/open-telemetry/opentelemetry-java/blob/master/sdk/src/main/java/io/opentelemetry/sdk/trace/config/TraceConfig.java. Basically `@AutoValue` will try to implement all the abstract methods in your class (usually prefixed with  `get` or `set`), then implement hashCode` and `equals` based on those fields.

For a complete guide, see https://github.com/google/auto/blob/master/value/userguide/howto.md.

Reply:
I don't think AutoValue is a good fit here.  Unless I'm missing something, I still am not sure how to make AutoValue work with this scenario.  Perhaps we can merge this and revisit later?

Reply:
SGTM, please leave a TODO here.